### DISCUSSION TITLE
Train the model using PyTorch's DDP mode, and when OOM occurs, the next trial will not proceed.

### DISCUSSION THREAD

**Discussion by Qianshaowei** [optuna/optuna#5191] [Q&A] [answered]
Thanks for your work. 
I train the model using PyTorch's DDP mode, and when OOM occurs, the next trial will not proceed. 
How should this problem be solved?

**Comment by nzw0301** [comment, marked as answer]
How about specifying `catch` argument of [`study.optimize`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.optimize) method?

**Reply by Qianshaowei** [reply]
if rank == 0:
    study = optuna.create_study(study_name = 'train',direction="maximize", sampler = optuna.samplers.TPESampler(), storage = 'sqlite:///db.sqlite3',load_if_exists=True)
    study.optimize(objective, n_trials=n_trials, show_progress_bar=True)
else:
    for _ in range(n_trials):
        try:
            objective(None)
        except optuna.TrialPruned:
            pass

**Reply by Qianshaowei** [reply]
I did not use parameters related to capturing exceptions. How to use catch to capture the out of memory exception  in DDP mode.

**Reply by nzw0301** [reply]
1. Did your script work even without the OOM? I suppose you need to use [`optuna.integration.TorchDistributedTrial`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.TorchDistributedTrial.html).
2. At least non-distributed case, an example of `catch` argument for OOM error would be 

```python
import torch
import optuna

def objective(trial):
    num_params = trial.suggest_int("x", 1, 10000)
    torch.rand(num_params**3, device="cuda") # to raise OOM
    return num_params

study = optuna.create_study()
study.optimize(objective, catch=(torch.cuda.OutOfMemoryError, ), n_trials=10)
```

I'm afraid that I don't have multiple GPUs to run the code with DDP.

**Reply by Qianshaowei** [reply]
I modified my code based on the following code. It can be trained normally, but due to the modification of the model structure in different trials, sometimes the model structure found during training may go out of memory. After this exception occurs, the program terminates running.
`https://github.com/optuna/optuna-examples/blob/main/pytorch/pytorch_distributed_simple.py`
Some error messages are as follows:
`"2023-10-24T11:18:44+08:00" Traceback (most recent call last):
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 587, in <module>
"2023-10-24T11:18:44+08:00"     objective(None)
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 420, in objective
"2023-10-24T11:18:44+08:00"     ds_loss = args.softmax_ratio * ds_net(all_feats, all_labels)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
"2023-10-24T11:18:44+08:00"     result = self.forward(*input, **kwargs)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 194, in forward
"2023-10-24T11:18:44+08:00"     self.device_stored_person_offset, self.margin, self.scale, self.normalize, self.neg_th, self.neg_margin)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 80, in forward
"2023-10-24T11:18:44+08:00"     curr_logits = torch.exp(curr_logits)
"2023-10-24T11:18:44+08:00" RuntimeError: CUDA out of memory. Tried to allocate 1.56 GiB (GPU 4; 31.75 GiB total capacity; 29.08 GiB already allocated; 659.75 MiB free; 29.86 GiB reserved in total by PyTorch)
"2023-10-24T11:18:44+08:00" Traceback (most recent call last):
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 587, in <module>
"2023-10-24T11:18:44+08:00"     objective(None)
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 420, in objective
"2023-10-24T11:18:44+08:00"     ds_loss = args.softmax_ratio * ds_net(all_feats, all_labels)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
"2023-10-24T11:18:44+08:00"     result = self.forward(*input, **kwargs)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 194, in forward
"2023-10-24T11:18:44+08:00"     self.device_stored_person_offset, self.margin, self.scale, self.normalize, self.neg_th, self.neg_margin)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 80, in forward
"2023-10-24T11:18:44+08:00"     curr_logits = torch.exp(curr_logits)
"2023-10-24T11:18:44+08:00" RuntimeError: CUDA out of memory. Tried to allocate 1.56 GiB (GPU 1; 31.75 GiB total capacity; 29.08 GiB already allocated; 659.75 MiB free; 29.86 GiB reserved in total by PyTorch)
"2023-10-24T11:18:44+08:00" Traceback (most recent call last):
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 587, in <module>
"2023-10-24T11:18:44+08:00"     objective(None)
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 420, in objective
"2023-10-24T11:18:44+08:00"     ds_loss = args.softmax_ratio * ds_net(all_feats, all_labels)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
"2023-10-24T11:18:44+08:00"     result = self.forward(*input, **kwargs)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 194, in forward
"2023-10-24T11:18:44+08:00"     self.device_stored_person_offset, self.margin, self.scale, self.normalize, self.neg_th, self.neg_margin)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 80, in forward
"2023-10-24T11:18:44+08:00"     curr_logits = torch.exp(curr_logits)
"2023-10-24T11:18:44+08:00" RuntimeError: CUDA out of memory. Tried to allocate 1.56 GiB (GPU 5; 31.75 GiB total capacity; 29.08 GiB already allocated; 659.75 MiB free; 29.86 GiB reserved in total by PyTorch)
"2023-10-24T11:18:44+08:00" Traceback (most recent call last):
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 587, in <module>
"2023-10-24T11:18:44+08:00"     objective(None)
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 420, in objective
"2023-10-24T11:18:44+08:00"     ds_loss = args.softmax_ratio * ds_net(all_feats, all_labels)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
"2023-10-24T11:18:44+08:00" 
                                                 
[W 2023-10-24 11:18:44,666] Trial 1 failed with parameters: {'layer_3': 5, 'layer_2': 15} because of the following error: RuntimeError('CUDA out of memory. Tried to allocate 1.56 GiB (GPU 0; 31.75 GiB total capacity; 29.08 GiB already allocated; 659.75 MiB free; 29.86 GiB reserved in total by PyTorch)').
"2023-10-24T11:18:44+08:00" Traceback (most recent call last):
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/optuna/study/_optimize.py", line 200, in _run_trial
"2023-10-24T11:18:44+08:00"     value_or_values = func(trial)
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 420, in objective
"2023-10-24T11:18:44+08:00"     ds_loss = args.softmax_ratio * ds_net(all_feats, all_labels)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
"2023-10-24T11:18:44+08:00"     result = self.forward(*input, **kwargs)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 194, in forward
"2023-10-24T11:18:44+08:00"     self.device_stored_person_offset, self.margin, self.scale, self.normalize, self.neg_th, self.neg_margin)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 80, in forward
"2023-10-24T11:18:44+08:00"     curr_logits = torch.exp(curr_logits)
"2023-10-24T11:18:44+08:00" RuntimeError: CUDA out of memory. Tried to allocate 1.56 GiB (GPU 0; 31.75 GiB total capacity; 29.08 GiB already allocated; 659.75 MiB free; 29.86 GiB reserved in total by PyTorch)
"2023-10-24T11:18:44+08:00" 
  2%|▎         | 1/40 [03:15<1:09:57, 107.63s/it]    
                                                 
result = self.forward(*input, **kwargs)
"2023-10-24T11:18:44+08:00" [W 2023-10-24 11:18:44,668] Trial 1 failed with value None.
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 194, in forward
"2023-10-24T11:18:44+08:00" 
  2%|▎         | 1/40 [03:15<1:09:57, 107.63s/it]
  2%|▎         | 1/40 [03:15<2:06:53, 195.22s/it]
"2023-10-24T11:18:44+08:00"     self.device_stored_person_offset, self.margin, self.scale, self.normalize, self.neg_th, self.neg_margin)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 80, in forward
"2023-10-24T11:18:44+08:00"     curr_logits = torch.exp(curr_logits)
"2023-10-24T11:18:44+08:00" RuntimeError: CUDA out of memory. Tried to allocate 1.56 GiB (GPU 7; 31.75 GiB total capacity; 29.08 GiB already allocated; 659.75 MiB free; 29.86 GiB reserved in total by PyTorch)
"2023-10-24T11:18:44+08:00" Traceback (most recent call last):
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 583, in <module>
"2023-10-24T11:18:44+08:00"     study.optimize(objective, n_trials=n_trials, show_progress_bar=True)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/optuna/study/study.py", line 460, in optimize
"2023-10-24T11:18:44+08:00"     show_progress_bar=show_progress_bar,
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/optuna/study/_optimize.py", line 76, in _optimize
"2023-10-24T11:18:44+08:00"     progress_bar=progress_bar,
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/optuna/study/_optimize.py", line 163, in _optimize_sequential
"2023-10-24T11:18:44+08:00"     frozen_trial = _run_trial(study, func, catch)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/optuna/study/_optimize.py", line 251, in _run_trial
"2023-10-24T11:18:44+08:00"     raise func_err
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/optuna/study/_optimize.py", line 200, in _run_trial
"2023-10-24T11:18:44+08:00"     value_or_values = func(trial)
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 420, in objective
"2023-10-24T11:18:44+08:00"     ds_loss = args.softmax_ratio * ds_net(all_feats, all_labels)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
"2023-10-24T11:18:44+08:00"     result = self.forward(*input, **kwargs)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 194, in forward
"2023-10-24T11:18:44+08:00"     self.device_stored_person_offset, self.margin, self.scale, self.normalize, self.neg_th, self.neg_margin)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 80, in forward
"2023-10-24T11:18:44+08:00"     curr_logits = torch.exp(curr_logits)
"2023-10-24T11:18:44+08:00" RuntimeError: CUDA out of memory. Tried to allocate 1.56 GiB (GPU 0; 31.75 GiB total capacity; 29.08 GiB already allocated; 659.75 MiB free; 29.86 GiB reserved in total by PyTorch)
"2023-10-24T11:18:44+08:00" Traceback (most recent call last):
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 587, in <module>
"2023-10-24T11:18:44+08:00"     objective(None)
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 420, in objective
"2023-10-24T11:18:44+08:00"     ds_loss = args.softmax_ratio * ds_net(all_feats, all_labels)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
"2023-10-24T11:18:44+08:00"     result = self.forward(*input, **kwargs)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 194, in forward
"2023-10-24T11:18:44+08:00"     self.device_stored_person_offset, self.margin, self.scale, self.normalize, self.neg_th, self.neg_margin)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 80, in forward
"2023-10-24T11:18:44+08:00"     curr_logits = torch.exp(curr_logits)
"2023-10-24T11:18:44+08:00" RuntimeError: CUDA out of memory. Tried to allocate 1.56 GiB (GPU 6; 31.75 GiB total capacity; 29.08 GiB already allocated; 659.75 MiB free; 29.86 GiB reserved in total by PyTorch)
"2023-10-24T11:18:44+08:00" Traceback (most recent call last):
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 587, in <module>
"2023-10-24T11:18:44+08:00"     objective(None)
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 420, in objective
"2023-10-24T11:18:44+08:00"     ds_loss = args.softmax_ratio * ds_net(all_feats, all_labels)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
"2023-10-24T11:18:44+08:00"     result = self.forward(*input, **kwargs)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 194, in forward
"2023-10-24T11:18:44+08:00"     self.device_stored_person_offset, self.margin, self.scale, self.normalize, self.neg_th, self.neg_margin)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 80, in forward
"2023-10-24T11:18:44+08:00"     curr_logits = torch.exp(curr_logits)
"2023-10-24T11:18:44+08:00" RuntimeError: CUDA out of memory. Tried to allocate 1.56 GiB (GPU 3; 31.75 GiB total capacity; 29.08 GiB already allocated; 659.75 MiB free; 29.86 GiB reserved in total by PyTorch)
"2023-10-24T11:18:44+08:00" Traceback (most recent call last):
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 587, in <module>
"2023-10-24T11:18:44+08:00"     objective(None)
"2023-10-24T11:18:44+08:00"   File "../auto_search.py", line 420, in objective
"2023-10-24T11:18:44+08:00"     ds_loss = args.softmax_ratio * ds_net(all_feats, all_labels)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 722, in _call_impl
"2023-10-24T11:18:44+08:00"     result = self.forward(*input, **kwargs)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 194, in forward
"2023-10-24T11:18:44+08:00"     self.device_stored_person_offset, self.margin, self.scale, self.normalize, self.neg_th, self.neg_margin)
"2023-10-24T11:18:44+08:00"   File "/data1/auto_search/auto_search_hyper_v0.4_8gpu_esv2_t/pyce/losses/ds_softmax.py", line 80, in forward
"2023-10-24T11:18:44+08:00"     curr_logits = torch.exp(curr_logits)
"2023-10-24T11:18:44+08:00" RuntimeError: CUDA out of memory. Tried to allocate 1.56 GiB (GPU 2; 31.75 GiB total capacity; 29.08 GiB already allocated; 659.75 MiB free; 29.86 GiB reserved in total by PyTorch)
"2023-10-24T11:18:44+08:00" Exception in thread Thread-4:
"2023-10-24T11:18:44+08:00" Traceback (most recent call last):
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/threading.py", line 917, in _bootstrap_inner
"2023-10-24T11:18:44+08:00"     self.run()
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/threading.py", line 865, in run
"2023-10-24T11:18:44+08:00"     self._target(*self._args, **self._kwargs)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/utils/data/_utils/pin_memory.py", line 25, in _pin_memory_loop
"2023-10-24T11:18:44+08:00"     r = in_queue.get(timeout=MP_STATUS_CHECK_INTERVAL)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/multiprocessing/queues.py", line 113, in get
"2023-10-24T11:18:44+08:00"     return _ForkingPickler.loads(res)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/multiprocessing/reductions.py", line 282, in rebuild_storage_fd
"2023-10-24T11:18:44+08:00"     fd = df.detach()
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/multiprocessing/resource_sharer.py", line 57, in detach
"2023-10-24T11:18:44+08:00"     with _resource_sharer.get_connection(self._id) as conn:
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/multiprocessing/resource_sharer.py", line 87, in get_connection
"2023-10-24T11:18:44+08:00"     c = Client(address, authkey=process.current_process().authkey)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/multiprocessing/connection.py", line 498, in Client
"2023-10-24T11:18:44+08:00"     answer_challenge(c, authkey)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/multiprocessing/connection.py", line 741, in answer_challenge
"2023-10-24T11:18:44+08:00"     message = connection.recv_bytes(256)         # reject large message
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/multiprocessing/connection.py", line 216, in recv_bytes
"2023-10-24T11:18:44+08:00"     buf = self._recv_bytes(maxlength)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/multiprocessing/connection.py", line 407, in _recv_bytes
"2023-10-24T11:18:44+08:00"     buf = self._recv(4)
"2023-10-24T11:18:44+08:00"   File "/usr/local/miniconda3/lib/python3.7/multiprocessing/connection.py", line 379, in _recv
"2023-10-24T11:18:44+08:00"     chunk = read(handle, remaining)
"2023-10-24T11:18:44+08:00" ConnectionResetError: [Errno 104] Connection reset by peer
"2023-10-24T11:18:44+08:00" 
"2023-10-24T11:18:49+08:00" Traceback (most recent call last):
"2023-10-24T11:18:49+08:00"   File "/usr/local/miniconda3/lib/python3.7/runpy.py", line 193, in _run_module_as_main
"2023-10-24T11:18:49+08:00"     "__main__", mod_spec)
"2023-10-24T11:18:49+08:00"   File "/usr/local/miniconda3/lib/python3.7/runpy.py", line 85, in _run_code
"2023-10-24T11:18:49+08:00"     exec(code, run_globals)
"2023-10-24T11:18:49+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/distributed/launch.py", line 261, in <module>
"2023-10-24T11:18:49+08:00"     main()
"2023-10-24T11:18:49+08:00"   File "/usr/local/miniconda3/lib/python3.7/site-packages/torch/distributed/launch.py", line 257, in main
"2023-10-24T11:18:49+08:00"     cmd=cmd)
"2023-10-24T11:18:49+08:00" subprocess.CalledProcessError: Command '['/usr/bin/python', '-u', '../auto_search.py', '--local_rank=7', '--nproc_per_node=8', '--idcard_in']' returned non-zero exit status 1.
"2023-10-24T11:18:49+08:00" *****************************************
"2023-10-24T11:18:49+08:00" Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
"2023-10-24T11:18:49+08:00" *****************************************`

**Reply by nzw0301** [reply]
Could you share the minimal reproducible code with us?

**Reply by Qianshaowei** [reply]
```
import optuna
from optuna.trial import TrialState
import os
import torch
import torch.distributed as dist
import torch.backends.cudnn as cudnn
import torch
import argparse

parser = argparse.ArgumentParser(description='PyTorch CIFAR10 Training')
""" Distributed """
parser.add_argument('--world_size', default=1, type=int,
                    help='number of distributed processes')
parser.add_argument('--dist_url', default='env://', type=str,
                    help='url used to set up distributed training')
parser.add_argument('--dist_backend', default='nccl', type=str,
                    help='distributed backend')
parser.add_argument("--local_rank", default=0, type=int,
                    help='local rank')
parser.add_argument("--global_rank", default=0, type=int,
                    help='global rank')
parser.add_argument("--nproc_per_node", type=int, default=1,
                    help="The number of processes to launch on each node, "
                         "for GPU training, this is recommended to be set "
                         "to the number of GPUs in your system so that "
                         "each process can be bound to a single GPU.")

args = parser.parse_args()
cudnn.benchmark = True
args.global_rank = int(os.environ['RANK'])
args.world_size = int(os.environ['WORLD_SIZE'])
device = torch.device('cuda:{}'.format(args.local_rank))
torch.cuda.set_device(args.local_rank)
dist.init_process_group(backend=args.dist_backend, init_method=args.dist_url)

def objective(single_trial):

    trial = optuna.integration.TorchDistributedTrial(single_trial)
    num_params = trial.suggest_int("x", 1, 10000)
    torch.rand(num_params**3, device=device) # to raise OOM

    if trial.should_prune():
        raise optuna.exceptions.TrialPruned()
    return num_params

if __name__ == "__main__":

    study = None
    n_trials = 40
    rank =args.global_rank
    if rank == 0:
        study = optuna.create_study(study_name = 'train',direction="maximize", sampler = optuna.samplers.TPESampler())
        study.optimize(objective, n_trials=n_trials, show_progress_bar=True)
    else:
        for _ in range(n_trials):
            try:
                objective(None)
            except optuna.TrialPruned:
                pass

    if rank == 0:
        assert study is not None
        pruned_trials = study.get_trials(deepcopy=False, states=[TrialState.PRUNED])
        complete_trials = study.get_trials(deepcopy=False, states=[TrialState.COMPLETE])

        print("Study statistics: ")
        print("  Number of finished trials: ", len(study.trials))
        print("  Number of pruned trials: ", len(pruned_trials))
        print("  Number of complete trials: ", len(complete_trials))

        print("Best trial:")
        trial = study.best_trial

        print("  Value: ", trial.value)

        print("  Params: ")
        for key, value in trial.params.items():
            print("    {}: {}".format(key, value))

#python -m torch.distributed.launch --nproc_per_node=${GPU_NUM} ${PYTHONSCRIPT} --nproc_per_node=${GPU_NUM}
```
The above code can reproduce this issue. To run this code, use command: 
`python -m torch.distributed.launch --nproc_per_node=${GPU_NUM} ${PYTHONSCRIPT} --nproc_per_node=${GPU_NUM}`

**Reply by nzw0301** [reply]
why not using https://github.com/optuna/optuna/discussions/5191#discussioncomment-8158480?

**Reply by Qianshaowei** [reply]
I have tried and the error message is `AttributeError：module ‘torch.cuda’ has no attribute 'OutOfMemoryError'`. I'm not sure if it's related to the Torch version, the version I'm using is `torch==1.6.0`.

**Reply by juancq** [reply]
I had a similar issue. I solved it by putting the line(s) that caused OOM errors in a try/except statement. Raise the optuna exception in the except statement.

**Reply by nzw0301** [reply]
Awesome. Thank you for sharing it!

**Reply by Qianshaowei** [reply]
Thank you for sharing it!
